04-frequent-scenarios.md

```diff
@@ -0,0 +1,288 @@
+# Frequent scenarios
+
+## Introduction
+
+Many use cases can be captured in a small set of scenarios. This chapter aims to illustrate this small set and get you up and running as soon as possible with the REST convention. 
+
+## Scenarios
+
+The scenarios described center around the [HTTP Response Status Codes][http-status-codes]. 
+
+We will use the following codes:
+* [200](#200--happy-flow); the expected response, ie the happy flow scenario
+* [301](#301--redirected); request for the client to retrieve the resource from another location
+* [304](#304--apply-cache); request for the client to utilize its cache
+* [400](#400--validation-error); an error was encountered that can be fixed by the client
+* [401](#401--unauthenticated); the client has not been authenticated
+* [403](#403--unauthorized); the client tried to access a resource for which it was not authorized
+* [404](#404--not-found); the requested resource could not be found
+* [500](#500--critical-error); the server encountered an error from which it cannot recover
+
+### 200 / Happy flow
+
+Typical response status code: 200
+
+Request
+```
+GET /cars/1
+```
+
+Response
+```
+Status code: 200
+Content: <JSON body of cars instance>
+```
+
+```java
+@RestController
+@RequestMapping("/cars")
+public class CarController {
+
+    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
+    public Car getCar(@PathVariable Long id) {
+        return carService.findOne(id);
+    }
+}
+```
+
+If the car resource is found, Spring will automatically make sure a response is created with a 200 code. You will have to do nothing special for this to work.
+
+### 301 / Redirected
+
+A redirection takes place if the server can determine whether a resource has been moved to another location, and also knows what that new location is. A typical example of valid usage here, is a resource which is addressed by a human-readable permalink (for example, the title of an article). When the title changes, the resource has effectively moved. This situation warrants a redirect operation.
+
+Ideally, if a resource has been moved to another link, the client will be told so in the standard HTTP way. In this case, the server method responsible for looking up the resource will throw an exception that is handled by a @ControllerAdvice method. This method will set the response code (301) and add the header with the new location of the resource.
+
+```java
+@ExceptionHandler(CarRedirectException.class)
```

Comment:
Particularly in this example, the exception would not have to be thrown in a controller method, but in a service method, since this is where the business logic gets handled.

We would like to use exceptions in the Java application to signal problems. The generic @ControllerAdvice will then deal with mapping the exception to a HTTP response. Having both this approach and specific response status creation in particular methods, mixes two different approach. I would say that picking one strategy trumps having both of them.

That said, there is a lot to be said for choosing a generic @ControllerAdvice to deal with mapping exceptions to specific HTTP responses. The application can up to the level of the afore-mentioned filter entirely work with exceptions and agnostic of the HTTP mapping. 

As to your question whether it's a standard way of dealing with 301 redirects in Spring MVC controllers; it is standard to redirect callers of no longer existing links to the new link. In our company, it is also a standard to delegate the translation from exceptions to HTTP responses to a particular @ControllerAdvice class. So, yeah, I guess so.